Content.Shared/Clothing/EntitySystems/ClothingSystem.cs

```diff
@@ -85,49 +85,77 @@ private void QuickEquip(
 
-            bool shouldLayerShow = true;
-            while (enumerator.NextItem(out EntityUid item, out SlotDefinition? slot))
+                    bool shouldLayerShow = true;
+                    SlotFlags hideSlot = hideLayer.Layers[layer];
+                    if (clothing.EquippedInSlot == hideSlot)
+                    {
+                        shouldLayerShow = MaskToggleCheck();
+                    }
+                    _humanoidSystem.SetLayerVisibility(equipee, layer, shouldLayerShow, hideSlot);
+                }
+            }
+            else
             {
-                if (TryComp(item, out HideLayerClothingComponent? comp))
+                // iterate the HideLayerClothingComponent's legacy slots set and check
+                // that the wearer has the layer to be hidden, and that the clothing
+                // is equipped.
+                foreach (HumanoidVisualLayers layer in hideLayer.Slots)
                 {
-                    if (comp.Slots.Contains(layer))
+                    if (!appearanceLayers.Contains(layer))
+                        continue;
+
+                    bool shouldLayerShow = true;
+                    if (clothing.EquippedInSlot != SlotFlags.NONE
+                        && clothing.Slots.HasFlag(clothing.EquippedInSlot))
                     {
-                        if (TryComp(item, out ClothingComponent? clothing) && clothing.Slots == slot.SlotFlags)
-                        {
-                            //Checks for mask toggling. TODO: Make a generic system for this
-                            if (comp.HideOnToggle && TryComp(item, out MaskComponent? mask))
-                            {
-                                if (clothing.EquippedPrefix != mask.EquippedPrefix)
-                                {
-                                    shouldLayerShow = false;
-                                    break;
-                                }
-                            }
-                            else
-                            {
-                                shouldLayerShow = false;
-                                break;
-                            }
-                        }
+                        shouldLayerShow = MaskToggleCheck();
                     }
+                    _humanoidSystem.SetLayerVisibility(equipee, layer, shouldLayerShow, clothing.Slots);
                 }
+
+            }
+        }
+
+        //Checks for mask toggling. TODO: Make a generic system for this
+        bool MaskToggleCheck()
+        {
+            if (hideLayer.HideOnToggle && TryComp(equipment, out MaskComponent? mask))
+            {
+                if (clothing.EquippedPrefix != mask.EquippedPrefix)
+                {
+                    return false;
+                }
+            }
+            else
+            {
+                return false;
             }
-            _humanoidSystem.SetLayerVisibility(equipee, layer, shouldLayerShow);
+            return true;
```

Comment:
IMO its nicer to read with less if/else nesting. I.e.,
```cs
bool MaskToggleCheck()
{
    if (!hideLayer.HideOnToggle || !TryComp(equipment, out MaskComponent? mask))
        return false;

    return clothing.EquippedPrefix == mask.EquippedPrefix;
}
```

Reply:
I can see what you mean. I think the double negative OR is a little harder to read, but I'm not particular about it.